src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,226 @@
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $serviceLocator = null
+    ) {
+    }
+
+    public function map(object $object, object|string $to = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $refl = new \ReflectionClass($object);
+
+        $metadata = $this->metadataFactory->create($object);
+        $map = $this->getMapTo($metadata, null, $object);
+        $to ??= $map?->to;
+
+        if (!(\is_string($to) && class_exists($to)) && !\is_object($to)) {
+            throw new RuntimeException(sprintf('Attribute of type "%s" expected on "%s.', Map::class, $refl->getName()));
+        }
+
+        $toRefl = new \ReflectionClass($to);
+        $mapped = \is_object($to) ? $to : $toRefl->newInstanceWithoutConstructor();
+        if ($map && $map->transform) {
+            $mapped = $this->applyTransforms($map, $mapped, $mapped);
+        }
+
+        if (!\is_object($mapped)) {
+            throw new RuntimeException('Can not map to a non-object.');
+        }
+
+        $objectMap[$object] = $mapped;
+
+        $arguments = [];
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            $property = $toRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mapped) && $property->getValue($mapped))) {
+                continue;
+            }
+
+            $arguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            if ($property->isStatic()) {
+                continue;
+            }
+
+            $propertyName = $property->getName();
+
+            $map = null;
+            foreach ($property->getAttributes(Map::class) as $attr) {
+                $map = $attr->newInstance();
+                $if = $map->if;
+
+                if (false === $if) {
+                    continue 2;
+                }
+
+                if ($if && ($fn = $this->getCallable($if)) && !$this->call($fn, null, $object)) {
+                    continue 2;
+                }
+
+                break;
+            }
+
+            $mapToProperty = $map?->to ?? $propertyName;
+            if (!$mapToProperty || !$toRefl->hasProperty($mapToProperty)) {
+                continue;
+            }
+
+            $value = $this->propertyAccessor ? $this->propertyAccessor->getValue($object, $propertyName) : $object->{$propertyName};
```

Comment:
What value does it give to use the property accessor ? Is it to be able to use a getter instead of property if available ?

I think it may be better to either : 

 * don't use property accessor here and only use direct access or reflection if private
 * or use `PropertyAccessor` + `PropertyInfo` to also extract properties from method
 
I fear that a mix between this 2 options would confuse users in regard to how other components works in symfony

Reply:
Why would we use `PropertyInfo` if we don't need type checking? PropertyAccessor is here because this will mostly be used with Doctrine and I don't want to reimplement the accessor logic.

Reply:
Mainly to use the `PropertyListExtractorInterface` to retrieve a list of properties since it works well when used in conjunction with `PropertyAcessor`, this will also allow to map from a method instead of only properties

Reply:
do you think we can keep that as a further improvement?

Reply:
Yes, i don't think this is a blocker for a first implementation

Reply:
@joelwurtz in my last comment the metadata layer is responsible to get the property mapping metadata, do you think: 
- I should add a metadata layer for property names (here `getProperties`)?
- Following the change that metadata and attributes would be splitted: is retrieving class mapping metadata and class property mapping metadata in the same class is fine or that I should split that as well? IMO the metadata is the same we want a source, target, ignore and transform callback.